### DISCUSSION TITLE
How to configure 'Result Cache' with no expiration? What about clearing all result cache entries by entity?

### DISCUSSION THREAD

**Discussion by Frank-D** [mikro-orm/mikro-orm#4490] [Q&A]
Hi,

**Q1: Unlimited / no expiration 'result cache'**
-
I was wondering if it's possible to configure the '[Result Cache](https://mikro-orm.io/docs/caching)' with no expiration for a given entity? 

**Context**:
The entity could represent some application reference data (e.g. data that is static most of the time, even though it is still dynamic data as it can be updated over time, just not frequently, hence I would like to avoid hitting the db as much as possible when fetching such entity; an example of this could be to hold application Feature Flags, or some application global configurations, etc, that can all be modified rarely from some ui admin panels..).

I tried using an expiration of '0' ms as in the below example, but it wasn't interpreted as 'unlimited expiration' but simply '0 ms expiration' (e.g. +- equivalent to disabling result caching with 'cache: false'...):
```
const res = await em.find(Book, { author: { name: 'Jon Snow' } }, {
  cache: ['book-cache-key', 0], // sets expiration to 0ms, seems almost similar to 'cache: false', in which case imo it would be more beneficial to make '0' behave as an 'unlimited time cache', since if I want to disable the cache I can already use 'cache: false'...
});
```

My objective with such unlimited cache support is to cache reference data as mentioned above, and manually manage myself 'when' such data becomes stale and requires to be manually reset/cleared (e.g. when an update or insert is made against that entity/table basically, simple..), using the provided method:
```
// clear the cache key by name
await em.clearCache('book-cache-key');
```

**Q2: Clearing all 'result cache' entries for a given entity**
-
Is it possible to clear all '[Result Cache](https://mikro-orm.io/docs/caching)' entries related to a specific entity?

Or perhaps I should ask first, if it's possible to "+- regroup in cache under the same entity" several different query results from different inputs but for all the same entity, so that later on we can easily clear all those specific entity cache entries in one go?

How could this be implemented? Well, perhaps if the cache could have 2 representations, one simple `[ cacheKey ] -> [ cacheValue]` map structure, and another one slightly more advanced which would be entity-based, e.g. `[ entity ] -> [ cacheKey ] -> [ cacheValue]`, basically implemented as a map of map.. This way, having the 2 cache representations internally, different cache utility methods could co-exist, some that work against the simple key->val structure (e.g. `em.clearCache('book-cache-key')`, while some others allow to deal with cache entries which are specific to an entity (e.g., `em.clearCacheEntries(Book)`).

---

 Right now, my 2 alternative options are:

1- Use extremely large 'expiration' result cache value, to simulate no expiration. 
(I haven't tested that option yet, not sure what's the maximum expiration ms value that is currently supported by such 'result cache' mechanism, but ideally if it could support expiration ms values large enough to hold cache data for multiple days or even weeks, that could be just fine to hit back the database every couple days/weeks to fetch back such unchanged application reference data and rebuild/reset the cache with those unchanged entries for another couple days/weeks..). Not ideal, but perhaps acceptable, the longer expiration supported the better to avoid unnecessary db queries.

2- Build my own caching layer at the service level, before hitting the mikro-orm repository / em.

Thanks

**Comment by B4nan** [comment]
I guess I don't mind doing the `0 = infinity`, but technically that's breaking, so it would go to v6, especially because this will require people to change their custom cache adapters most probably. 

The additional map for caching could work as well, or maybe we could introduce some cache tags, as that could be even more universal. Would you be up for a PR?
